cmd/systray/systray.go

```diff
@@ -47,26 +45,67 @@ var (
 )
 
 func main() {
-	systray.Run(onReady, onExit)
+	menu := new(Menu)
+	menu.updateState()
+
+	// exit cleanly on SIGINT and SIGTERM
+	go func() {
+		interrupt := make(chan os.Signal, 1)
+		signal.Notify(interrupt, syscall.SIGINT, syscall.SIGTERM)
+		select {
+		case <-interrupt:
+			menu.onExit()
+		case <-menu.bgCtx.Done():
+		}
+	}()
+
+	systray.Run(menu.onReady, menu.onExit)
 }
 
 // Menu represents the systray menu, its items, and the current Tailscale state.
 type Menu struct {
-	mu     sync.Mutex // protects the entire Menu
-	status *ipnstate.Status
+	mu sync.Mutex // protects the entire Menu
 
+	initOnce sync.Once
+
+	lc          tailscale.LocalClient
+	status      *ipnstate.Status
+	curProfile  ipn.LoginProfile
+	allProfiles []ipn.LoginProfile
+
+	bgCtx    context.Context // ctx for background tasks not involving menu item clicks
+	bgCancel context.CancelFunc
+
+	// Top-level menu items
 	connect    *systray.MenuItem
 	disconnect *systray.MenuItem
+	self       *systray.MenuItem
+	exitNodes  *systray.MenuItem
+	more       *systray.MenuItem
+	quit       *systray.MenuItem
 
-	self      *systray.MenuItem
-	more      *systray.MenuItem
-	exitNodes *systray.MenuItem
-	quit      *systray.MenuItem
-
+	rebuildCh  chan struct{} // triggers a menu rebuild
 	accountsCh chan ipn.ProfileID
 	exitNodeCh chan tailcfg.StableNodeID // ID of selected exit node
 
-	eventCancel func() // cancel eventLoop
+	eventCancel context.CancelFunc // cancel eventLoop
+
+	notificationIcon *os.File // icon used for desktop notifications
+}
+
+func (menu *Menu) init() {
+	menu.initOnce.Do(func() {
```

Comment:
if you're always calling init with menu.mu held, the sync.Once is kinda distracting/pointless? you could just return early if something's non-nil? (or add a `didInit bool`?)

But also fine, even if misleading a bit.

Reply:
done